File: 46. 全排列/Solution.py
class Solution:
    # 排列问题不需要考虑起始位置即不需要start了，而且需要每次都从开头寻找没有插入的数，使用used数组记录插入过的数
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        path = []
        used = [False] * len(nums)
        self.backtracking(ans, path, nums, used)
        return ans
    def backtracking(self, ans, path, nums, used):
        if len(path) == len(nums):
            ans.append(path[:])
            return
        
        for i in range(0, len(nums)):
            if used[i]:
                continue
            used[i] = True
            path.append(nums[i])
            self.backtracking(ans, path, nums, used)
            path.pop()
            used[i] = False

File: 46. 全排列/Solution.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> ans;
    vector<int> path;
    
    vector<vector<int>> permute(vector<int>& nums) {
        ans.clear();
        path.clear();
        vector<bool> used(nums.size(), false);
        backtracking(nums, used);
        return ans;
    }
    void backtracking(vector<int>& nums, vector<bool>& used){
        if(nums.size() == path.size()){
            ans.push_back(path);
            return;
        }
        for(int i = 0; i < nums.size(); i++){
            if(used[i] != false){
                continue;
            }
            used[i] = true;
            path.push_back(nums[i]);
            backtracking(nums, used);
            path.pop_back();
            used[i] = false;
        }
    }
};

